File: tracker/merger.py
import json
from datetime import datetime

from tracker.categorizer import categorize
from tracker.db import get_raw_events_for_date, delete_work_blocks_for_date, insert_work_block


def _parse_dt(dt_str: str) -> datetime:
    return datetime.strptime(dt_str, "%Y-%m-%dT%H:%M:%S")


def _gap_seconds(end_str: str, start_str: str) -> int:
    return int((_parse_dt(start_str) - _parse_dt(end_str)).total_seconds())


def merge_events_for_date(date: str) -> list[dict]:
    raw_events = get_raw_events_for_date(date)
    if not raw_events:
        return []

    # Step 1: Categorize and filter micro-sessions (< 2 min)
    categorized = []
    for event in raw_events:
        dur = event.get("duration_sec") or 0
        if dur < 120:
            continue
        category = categorize(event["app_name"], event.get("window_title"))
        categorized.append({
            "app_name": event["app_name"],
            "window_title": event.get("window_title"),
            "started_at": event["started_at"],
            "ended_at": event["ended_at"],
            "duration_sec": dur,
            "category": category,
        })

    if not categorized:
        return []

    # Step 2: Merge same-category interruptions (gap < 5 min)
    merged = [categorized[0].copy()]
    merged[0]["apps"] = {categorized[0]["app_name"]}

    for event in categorized[1:]:
        prev = merged[-1]
        gap = _gap_seconds(prev["ended_at"], event["started_at"])

        if event["category"] == prev["category"] and gap <= 300:
            prev["ended_at"] = event["ended_at"]
            prev["duration_sec"] = int((_parse_dt(prev["ended_at"]) - _parse_dt(prev["started_at"])).total_seconds())
            prev["apps"].add(event["app_name"])
        else:
            new_block = event.copy()
            new_block["apps"] = {event["app_name"]}
            merged.append(new_block)

    # Step 3: Absorb short switches (< 5 min sandwiched between same category)
    absorbed = list(merged)
    changed = True
    while changed:
        changed = False
        new_list = []
        i = 0
        while i < len(absorbed):
            if (
                i > 0
                and i < len(absorbed) - 1
                and absorbed[i]["duration_sec"] < 300
                and new_list
                and new_list[-1]["category"] == absorbed[i + 1]["category"]
            ):
                # Absorb middle into previous block
                new_list[-1]["ended_at"] = absorbed[i + 1]["ended_at"]
                new_list[-1]["duration_sec"] = int(
                    (_parse_dt(new_list[-1]["ended_at"]) - _parse_dt(new_list[-1]["started_at"])).total_seconds()
                )
                new_list[-1]["apps"].update(absorbed[i]["apps"])
                new_list[-1]["apps"].update(absorbed[i + 1]["apps"])
                i += 2
                changed = True
            else:
                new_list.append(absorbed[i])
                i += 1
        absorbed = new_list

    # Step 4: Filter blocks under 10 minutes
    final_blocks = [b for b in absorbed if b["duration_sec"] >= 600]

    # Step 5: Write to DB (idempotent — delete first)
    delete_work_blocks_for_date(date)

    result = []
    for block in final_blocks:
        duration_min = block["duration_sec"] // 60
        apps_used = json.dumps(sorted(block["apps"]))
        row_id = insert_work_block(
            date=date,
            started_at=block["started_at"],
            ended_at=block["ended_at"],
            duration_min=duration_min,
            category=block["category"],
            auto_category=block["category"],
            apps_used=apps_used,
            note=None,
        )
        result.append({
            "id": row_id,
            "date": date,
            "started_at": block["started_at"],
            "ended_at": block["ended_at"],
            "duration_min": duration_min,
            "category": block["category"],
            "auto_category": block["category"],
            "apps_used": apps_used,
            "note": None,
        })

    return result

File: electron/merger.ts
// electron/merger.ts
import { categorize } from "./categorizer";
import {
  getRawEventsForDate,
  deleteWorkBlocksForDate,
  insertWorkBlock,
} from "./db";

function parseDt(dtStr: string): Date {
  return new Date(dtStr); // ISO 8601 with T separator is well-defined in ECMAScript
}

function gapSeconds(endStr: string, startStr: string): number {
  return (parseDt(startStr).getTime() - parseDt(endStr).getTime()) / 1000;
}

function durationSeconds(startStr: string, endStr: string): number {
  return (parseDt(endStr).getTime() - parseDt(startStr).getTime()) / 1000;
}

interface MergeBlock {
  app_name: string;
  window_title: string | null;
  started_at: string;
  ended_at: string;
  duration_sec: number;
  category: string;
  apps: Set<string>;
}

export function mergeEventsForDate(date: string): any[] {
  const rawEvents = getRawEventsForDate(date);
  if (rawEvents.length === 0) return [];

  // Step 1: Categorize and filter micro-sessions (< 2 min)
  const categorized: MergeBlock[] = [];
  for (const event of rawEvents) {
    const dur = event.duration_sec ?? 0;
    if (dur < 120) continue;
    const category = categorize(event.app_name, event.window_title);
    categorized.push({
      app_name: event.app_name,
      window_title: event.window_title,
      started_at: event.started_at,
      ended_at: event.ended_at,
      duration_sec: dur,
      category,
      apps: new Set([event.app_name]),
    });
  }

  if (categorized.length === 0) return [];

  // Step 2: Merge same-category interruptions (gap < 5 min)
  const merged: MergeBlock[] = [{ ...categorized[0], apps: new Set(categorized[0].apps) }];

  for (let i = 1; i < categorized.length; i++) {
    const event = categorized[i];
    const prev = merged[merged.length - 1];
    const gap = gapSeconds(prev.ended_at, event.started_at);

    if (event.category === prev.category && gap <= 300) {
      prev.ended_at = event.ended_at;
      prev.duration_sec = durationSeconds(prev.started_at, prev.ended_at);
      event.apps.forEach((a) => prev.apps.add(a));
    } else {
      merged.push({ ...event, apps: new Set(event.apps) });
    }
  }

  // Step 3: Absorb short switches (< 5 min sandwiched between same category)
  let absorbed = [...merged];
  let changed = true;
  while (changed) {
    changed = false;
    const newList: MergeBlock[] = [];
    let i = 0;
    while (i < absorbed.length) {
      if (
        i > 0 &&
        i < absorbed.length - 1 &&
        absorbed[i].duration_sec < 300 &&
        newList.length > 0 &&
        newList[newList.length - 1].category === absorbed[i + 1].category
      ) {
        const prev = newList[newList.length - 1];
        prev.ended_at = absorbed[i + 1].ended_at;
        prev.duration_sec = durationSeconds(prev.started_at, prev.ended_at);
        absorbed[i].apps.forEach((a) => prev.apps.add(a));
        absorbed[i + 1].apps.forEach((a) => prev.apps.add(a));
        i += 2;
        changed = true;
      } else {
        newList.push(absorbed[i]);
        i += 1;
      }
    }
    absorbed = newList;
  }

  // Step 4: Filter blocks under 10 minutes
  const finalBlocks = absorbed.filter((b) => b.duration_sec >= 600);

  // Step 5: Write to DB (idempotent — delete first)
  deleteWorkBlocksForDate(date);

  const result: any[] = [];
  for (const block of finalBlocks) {
    const durationMin = Math.floor(block.duration_sec / 60);
    const appsUsed = JSON.stringify([...block.apps].sort());
    const rowId = insertWorkBlock({
      date,
      started_at: block.started_at,
      ended_at: block.ended_at,
      duration_min: durationMin,
      category: block.category,
      auto_category: block.category,
      apps_used: appsUsed,
      note: null,
    });
    result.push({
      id: rowId,
      date,
      started_at: block.started_at,
      ended_at: block.ended_at,
      duration_min: durationMin,
      category: block.category,
      auto_category: block.category,
      apps_used: appsUsed,
      note: null,
    });
  }

  return result;
}
